### PR TITLE
chore(widgets): Implement `Debug & Default` common traits

### PR DESCRIPTION
**Pull request by TieWay59**
Implement `Debug & Default` common traits for most structs in src.

Reorder the derive fields to be more consistent:

    Debug, Default, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash

see: https://github.com/ratatui-org/ratatui/issues/307

### CODE REVIEW

**Comment by TieWay59** [comment]
Hey @joshka, I've been doing this today. And I wish to share my discoveries with you before more next step of review. 

1. I can't split these files into four PR, because many of them are related (for example, if A deps on B, then we need B to have some C-Trait, then, add it to A.) So this is a huge one hard to review, if you have any idea on splitting this up, let me know if possible.
2. I following an intention to give as many common traits as we can, in  order of:
    - [Copy](https://doc.rust-lang.org/std/marker/trait.Copy.html)
    - [Clone](https://doc.rust-lang.org/std/clone/trait.lone.html)
    - [Eq](https://doc.rust-lang.org/std/cmp/trait.Eq.html)
    - [PartialEq](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html)
    - [Ord](https://doc.rust-lang.org/std/cmp/trait.Ord.html)
    - [PartialOrd](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html)
    - [Hash](https://doc.rust-lang.org/std/hash/trait.Hash.html) 
    - [Debug](https://doc.rust-lang.org/std/fmt/trait.Debug.html)
    - [Display](https://doc.rust-lang.org/std/fmt/trait.Display.html)
    - [Default](https://doc.rust-lang.org/std/default/trait.Default.html) 
3. But.. `Copy` trait might be implicitly memory-costing if have it around too much. 
4. `Display` is more difficult to design for every struct, so I never actively add it.
5. Some `Default impl` can be replaced by derive if every member has a `Default` trait already, I tried some cases, and I wonder if it's necessary to write a test case for it. 
6. `Ord`, and `PartialOrd`, might seem redundant but I add them all the way if it's possible.

File: src/buffer.rs
Comment (joshka) [inline src/buffer.rs:17]: Nit: can we make the order of the derives the same. This is mostly because I'm a little OCD about things that are the same should look the same, but it does have the benefit of being really easy to see at a glance which things are supported if the traits are always in the same order. Perhaps use the following order: Debug, Default, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash.

Taking a look at Ord/PartialOrd and the things which implement these, perhaps it doesn't make sense for most of the types to implement this. Perhaps let's defer that?

File: src/layout.rs
Comment (joshka) [inline src/layout.rs:42]: I wonder if Constraint makes sense to have a Default. Where would you see this being used?

File: src/layout.rs
Comment (joshka) [inline src/layout.rs:24]: We actually use Vertical as the default layout direction in Layout::new()

File: src/symbols.rs
Comment (joshka) [inline src/symbols.rs:26]: NINE_LEVELS is probably more generally useful for these as it shows more information to the user whereas three levels throws some away.

File: src/symbols.rs
Comment (joshka) [inline src/symbols.rs:80]: Same - nine levels?

File: src/widgets/gauge.rs
Comment (joshka) [inline src/widgets/gauge.rs:368]: The naming of this implies that there is something interesting about the old defaults, but it's not obvious what that might be.

Try to phrase tests as defining current behavior unless there's something specifically novel about what the old behavior was (and if that's so, add docs on why that is interesting). Here I think the defaults are just ordinary defaults and `fn default()` would be fine with `assert_eq!(LineGauge::default(), LineGauged { ... }`

**Review by joshka — CHANGES_REQUESTED** [review]
LGTM generally - I'll answer the question comments in the conversation

My main suggestion is to remove the Eq/PartialEq where it doesn't seem to make sense (unless you think we need it for some reason).

**Comment by joshka** [comment]
> Hey @joshka, I've been doing this today. And I wish to share my discoveries with you before more next step of review.

Nice

>     1. I can't split these files into four PR, because many of them are related (for example, if A deps on B, then we need B to have some C-Trait, then, add it to A.) So this is a huge one hard to review, if you have any idea on splitting this up, let me know if possible.

Makes sense, I'd even prefer to see a single commit if the split commits don't compile (as it makes it easier to `git bisect` if every commit compiles). No big deal on this as the merge will probably join the commits, but doesn't hurt.

>     2. I following an intention to give as many common traits as we can, in  order of:
>        
>        * [Copy](https://doc.rust-lang.org/std/marker/trait.Copy.html)
>        * [Clone](https://doc.rust-lang.org/std/clone/trait.lone.html)
>        * [Eq](https://doc.rust-lang.org/std/cmp/trait.Eq.html)
>        * [PartialEq](https://doc.rust-lang.org/std/cmp/trait.PartialEq.html)
>        * [Ord](https://doc.rust-lang.org/std/cmp/trait.Ord.html)
>        * [PartialOrd](https://doc.rust-lang.org/std/cmp/trait.PartialOrd.html)
>        * [Hash](https://doc.rust-lang.org/std/hash/trait.Hash.html)
>        * [Debug](https://doc.rust-lang.org/std/fmt/trait.Debug.html)
>        * [Display](https://doc.rust-lang.org/std/fmt/trait.Display.html)
>        * [Default](https://doc.rust-lang.org/std/default/trait.Default.html)

Note my comment about ordering. Of these, I'd probably say that Debug and Default are the two most important, followed by Clone and Copy. Being able to glance at a derives list and see that these are implemented is something I personally often rely on. Small nit but useful I think.

>     3. But.. `Copy` trait might be implicitly memory-costing if have it around too much.

Can you point out the impact of any problems you might see with this?

>     4. `Display` is more difficult to design for every struct, so I never actively add it.

Let's defer Display to another PR?

>     5. Some `Default impl` can be replaced by derive if every member has a `Default` trait already, I tried some cases, and I wonder if it's necessary to write a test case for it.

It's probably worthwhile adding default tests as it helps catch any future changes in the defaults that are accidentally made.

>     6. `Ord`, and `PartialOrd`, might seem redundant but I add them all the way if it's possible.

Perhaps these should not be implemented except for where it's necessary? Is there better guidance on this available? Are there use cases where having these makes things easier to implement that aren't obvious?

File: src/layout.rs
Comment (TieWay59) [reply src/layout.rs:42]: I just saw this is more widely used, but it's OK to remove it if not so helpful.

File: src/symbols.rs
Comment (TieWay59) [reply src/symbols.rs:26]: OK, learned it today~

File: src/widgets/gauge.rs
Comment (TieWay59) [reply src/widgets/gauge.rs]: I got your point, but still have no idea what test name is better, what do you like?

File: src/buffer.rs
Comment (TieWay59) [reply src/buffer.rs]: I agree~

**Comment by TieWay59** [comment]
@joshka Now my PR is focused on Debug & Default trait, I hope this shift can make you feel better while reviewing. And I also reordered some old patterns to align with the preferred traits order.

File: src/widgets/gauge.rs
Comment (joshka) [reply src/widgets/gauge.rs]: I'd go with just `fn default()` as the test name.

File: src/widgets/gauge.rs
Comment (joshka) [inline src/widgets/gauge.rs:370]: Is there a blocker to implementing these (or is it just that doing so means a bunch of changes to all the types?)

**Review by joshka — APPROVED** [review]
LGTM - just one question that is not blocking.

File: src/widgets/gauge.rs
Comment (TieWay59) [reply src/widgets/gauge.rs:370]:
```rust
#[derive(Debug, Default, PartialEq)] // <-- if add `PartialEq`
pub struct LineGauge<'a> {
    block: Option<Block<'a>>,
    ratio: f64,
    label: Option<Line<'a>>,
    line_set: symbols::line::Set, // <-- we'll need this to have PartialEq trait
    style: Style,
    gauge_style: Style,
}
```

So, I suppose deferring it would be nice.

**Comment by TieWay59** [comment]
@joshka About the coverage, Should I write a test for every struct? I found it would be very huge, and since there are still many old codes that don't have coverage before me, I'd like to hear your idea. 

It's OK if you feel this is necessary. But I have to warn that, since we don't have much `PartialEq` trait, I may have to write cases like:

```rust
        assert_eq!(
            format!("{:?}", LineGauge::default()),
            format!(
                "{:?}",
                LineGauge {
                    block: None,
                    ratio: 0.0,
                    label: None,
                    style: Style::default(),
                    line_set: symbols::line::NORMAL,
                    gauge_style: Style::default(),
                }
            ),
            "LineGauge::default() should have correct default values."
        );
```

(Yes, it looks weird. it's my workaround, please guide me if you have a better suggestion :heart: )

**Comment by TieWay59** [comment]
Sorry I pushed again, I found one struct I missed, and I give a shorter name for the default test.

**Review by joshka — APPROVED** [review]

**Comment by joshka** [comment]
I'd say that adding coverage on the derived traits can probably be deferred. I'll leave that up for whoever does the second review on this.

**Comment by kdheepak** [comment]
I don't have a blocking review, but I was curious what the impact on library size (or binary size) this has. Is there some metrics you can add to the PR description?

**Review by kdheepak — APPROVED** [review]

**Comment by joshka** [comment]
> I don't have a blocking review, but I was curious what the impact on library size (or binary size) this has. Is there some metrics you can add to the PR description?

This is pretty much zero.
I check on the debug build one of the examples shows up in the debug builds as 4KB extra on a 3.8MB binary. The release mode build is the exact same size. I'd assume most use cases will never care about binary size for this sort of thing, but if there's something obvious I'm missing...

**Comment by joshka** [comment]
The scrollbar::Set type moved to symbols::scrollbar::Set. I updated the commit with that change (just needs an approval to merge).

Thanks for submitting this PR. It's something that I have run into a few times (particularly Debug being missing).

**Comment by joshka** [comment]
P.s. to make future PRs easier for us merge, can you please setup commit signing?
See https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits

**Pull request merged**
